language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -1 +1,1798 @@
+    LValue, LValue_, parse_field, Field_, TypeVar_, Function_, StructDefinition_,
+};
+use crate::lexer::*;
+use libra_types::{account_address::AccountAddress, byte_array::ByteArray};
+use hex;
+
+// FIXME: The following simplified version of ParseError copied from
+// lalrpop-util should be replaced.
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ParseError<L, T, E> {
+    InvalidToken { location: L },
+    UnrecognizedToken {
+        token: (L, T, L),
+        expected: Vec<String>,
+    },
+    User { error: E },
+}
+
+/// Format a list of expected tokens.
+fn fmt_expected(f: &mut fmt::Formatter, expected: &[String]) -> fmt::Result {
+    if !expected.is_empty() {
+        writeln!(f)?;
+        for (i, e) in expected.iter().enumerate() {
+            let sep = match i {
+                0 => "Expected one of",
+                _ if i < expected.len() - 1 => ",",
+                // Last expected message to be written
+                _ => " or",
+            };
+            write!(f, "{} {}", sep, e)?;
+        }
+    }
+    Ok(())
+}
+
+impl<L, T, E> fmt::Display for ParseError<L, T, E>
+where
+    L: fmt::Display,
+    T: fmt::Display,
+    E: fmt::Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::ParseError::*;
+        match *self {
+            User { ref error } => write!(f, "{}", error),
+            InvalidToken { ref location } => write!(f, "Invalid token at {}",
+            location),
+            UnrecognizedToken {
+                token: (ref start, ref token, ref end),
+                ref expected,
+            } => {
+                write!(
+                    f,
+                    "Unrecognized token `{}` found at {}:{}",
+                    token, start, end
+                )?;
+                fmt_expected(f, expected)
+            }
+        }
+    }
+}
+
+fn consume_token<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>, tok: Tok)
+    -> Result<(), ParseError<usize, Token<'input>, failure::Error>>
+{
+    if (tokens.token.1).0 != tok {
+        return Err(ParseError::InvalidToken { location: tokens.token.0 });
+    }
+    tokens.advance()?;
+    Ok(())
+}
+
+fn parse_name<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<String, ParseError<usize, Token<'input>, failure::Error>>
+{
+    if (tokens.token.1).0 != Tok::NameValue {
+        return Err(ParseError::InvalidToken { location: tokens.token.0 });
+    }
+    let name = (tokens.token.1).1.to_string();
+    tokens.advance()?;
+    Ok(name)
+}
+
+fn parse_name_begin_ty<'input, 'builder>(tokens: &mut __Matcher<'input, 'builder>)
+    -> Result<String, ParseError<usize, Token<'input>, failure::Error>>
+{
+    if (tokens.token.1).0 != Tok::NameBeginTyValue {
+        return Err(ParseError::InvalidToken { location: tokens.token.0 });
+    }
+    let s = (tokens.token.1).1;
+    tokens.advance()?;
```

Comment:
Should we wrap all of these in accessors?

So we would have 
```
 if  tokens.peek() != Tok::NameBeginTyValue {
   return Err(ParseError::InvalidToken { location: tokens.token.0 });
 }
 let s = tokens.pop()?;
```

Reply:
OK. I was going to do that in a follow up, but I went ahead and did it now. I considered a pop() function, but decided to skip that for now, since there aren't that many places that would use it. The lexer functions are now: peek, content, start_loc, previous_end_loc, advance. (The previous_end_loc function is new and fixes a problem with trailing whitespace being included in the span ranges.)